daml-assistant/daml-helper/src/DamlHelper.hs

```diff
@@ -126,18 +130,227 @@ withJar jarPath args a = do
+
+    -- cases 5 or 6
+    daProjectRootM <- findDaProjectRoot targetFolderAbs
+    whenJust daProjectRootM $ \projectRoot -> do
+        when (targetFolderAbs /= projectRoot) $ do
+            let projectRootRel = makeRelative currentDir projectRoot
+            hPutStr stderr $ unlines
+                [ "ERROR: daml init target is not DA project root."
+                , "    daml init target  = " <> targetFolder
+                , "    DA project root   = " <> projectRootRel
+                , ""
+                , "To proceed with da.yaml migration, please use the project root:"
+                , "    daml init " <> escapePath projectRootRel
+                ]
+            exitFailure
+
+        let legacyConfigPath = projectRoot </> legacyConfigName
+            legacyConfigRel = normalise (targetFolderRel </> legacyConfigName)
+              -- ^ for display purposes
+
+        daYaml <- requiredE ("Failed to parse " <> T.pack legacyConfigPath) =<<
+            Y.decodeFileEither (projectRoot </> legacyConfigName)
+
+
+        putStr $ unlines
+            [ "Detected DA project."
+            , "Migrating " <> legacyConfigRel <> " to " <> projectConfigRel
+            ]
+
+
+        let getField :: Y.FromJSON t => T.Text -> IO t
+            getField name =
+                required ("Failed to parse project." <> name <> " from " <> T.pack legacyConfigPath) $
+                    flip Y.parseMaybe daYaml $ \y -> do
+                        p <- y Y..: "project"
+                        p Y..: name
+
+        minimumSdkVersion <- getMinimumSdkVersion
+        projSdkVersion :: SdkVersion <- getField "sdk-version"
+        let newProjSdkVersion = max projSdkVersion minimumSdkVersion
+
+        when (projSdkVersion < minimumSdkVersion) $ do
+            putStr $ unlines
+                [ ""
+                , "WARNING: da.yaml SDK version " <> versionToString projSdkVersion <> " is too old for the new"
+                , "assistant, so daml.yaml will use SDK version " <> versionToString newProjSdkVersion <> " instead."
+                , ""
+                ]
+
+        projSource :: T.Text <- getField "source"
+        projParties :: [T.Text] <- getField "parties"
+        projName :: T.Text <- getField "name"
+        projScenario :: T.Text <- getField "scenario"
+
+        BS.writeFile (projectRoot </> projectConfigName) . Y.encodePretty yamlConfig $ Y.object
+            [ ("sdk-version", Y.String (versionToText newProjSdkVersion))
+            , ("name", Y.String projName)
+            , ("source", Y.String projSource)
+            , ("scenario", Y.String projScenario)
+            , ("parties", Y.array (map Y.String projParties))
+            , ("version", Y.String "1.0.0")
+            , ("exposed-modules", Y.array [Y.String "Main"])
+            , ("dependencies", Y.array [Y.String "daml-prim", Y.String "daml-stdlib"])
+            ]
+
+
+        putStrLn ("Done! Please verify " <> projectConfigRel)
+        exitSuccess
```

Comment:
Not sure how I feel about `exitSuccess` as a `return` - feels a bit like exceptions as control flow... I can live with it though, the code is very neat.